File: src/main/java/org/seancheer/linkedlist/_160IntersectionOfTwoLinkedLists.java
package org.seancheer.linkedlist;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;

/**
 * Given the heads of two singly linked-lists headA and headB, return the node at which the two lists intersect. If the two linked lists have no intersection at all, return null.
 * <p>
 * For example, the following two linked lists begin to intersect at node c1:
 * <p>
 * <p>
 * The test cases are generated such that there are no cycles anywhere in the entire linked structure.
 * <p>
 * Note that the linked lists must retain their original structure after the function returns.
 * <p>
 * Custom Judge:
 * <p>
 * The inputs to the judge are given as follows (your program is not given these inputs):
 * <p>
 * intersectVal - The value of the node where the intersection occurs. This is 0 if there is no intersected node.
 * listA - The first linked list.
 * listB - The second linked list.
 * skipA - The number of nodes to skip ahead in listA (starting from the head) to get to the intersected node.
 * skipB - The number of nodes to skip ahead in listB (starting from the head) to get to the intersected node.
 * The judge will then create the linked structure based on these inputs and pass the two heads, headA and headB to your program. If you correctly return the intersected node, then your solution will be accepted.
 * <p>
 * <p>
 * <p>
 * Example 1:
 * <p>
 * <p>
 * Input: intersectVal = 8, listA = [4,1,8,4,5], listB = [5,6,1,8,4,5], skipA = 2, skipB = 3
 * Output: Intersected at '8'
 * Explanation: The intersected node's value is 8 (note that this must not be 0 if the two lists intersect).
 * From the head of A, it reads as [4,1,8,4,5]. From the head of B, it reads as [5,6,1,8,4,5]. There are 2 nodes before the intersected node in A; There are 3 nodes before the intersected node in B.
 * Example 2:
 * <p>
 * <p>
 * Input: intersectVal = 2, listA = [1,9,1,2,4], listB = [3,2,4], skipA = 3, skipB = 1
 * Output: Intersected at '2'
 * Explanation: The intersected node's value is 2 (note that this must not be 0 if the two lists intersect).
 * From the head of A, it reads as [1,9,1,2,4]. From the head of B, it reads as [3,2,4]. There are 3 nodes before the intersected node in A; There are 1 node before the intersected node in B.
 * Example 3:
 * <p>
 * <p>
 * Input: intersectVal = 0, listA = [2,6,4], listB = [1,5], skipA = 3, skipB = 2
 * Output: No intersection
 * Explanation: From the head of A, it reads as [2,6,4]. From the head of B, it reads as [1,5]. Since the two lists do not intersect, intersectVal must be 0, while skipA and skipB can be arbitrary values.
 * Explanation: The two lists do not intersect, so return null.
 * 保证不会出现环，要求O(n)和O(1)解决
 *
 * @author: seancheer
 * @date: 2021/8/10
 **/
public class _160IntersectionOfTwoLinkedLists extends LinkedListParent {

    public static void main(String[] args) {
        //其他用例直接去对应题目里面测试，因为这个测试场景比较复杂，不好构造
        var obj = new _160IntersectionOfTwoLinkedLists();
        String l1 = "[1,2,3,4]";
        var head = convertArrToListNode(converStrTo1DemArr(l1));
        var head2 = convertArrToListNode(converStrTo1DemArr(l1));
        var res = obj.getIntersectionNode(head, head2);
        var res2 = obj.getIntersectionNode2(head, head2);
        System.out.println("res:" + toStrListNode(res) + "   res2:" + toStrListNode(res2));//null
    }

    /**
     * 解题思路：比较直观的思路，把链表分别压到两个栈，然后两个栈依次弹栈，直到栈顶出现不一样的node的时候，就是最终的答案，
     * 但是这种解法的空间复杂度不是O(1)
     *
     * @param headA
     * @param headB
     * @return
     */
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if (null == headA || null == headB) {
            return null;
        }

        Deque<ListNode> stackA = new ArrayDeque<>();
        Deque<ListNode> stackB = new ArrayDeque<>();
        ListNode cur = headA;
        while (null != cur) {
            stackA.push(cur);
            cur = cur.next;
        }
        cur = headB;
        while (null != cur) {
            stackB.push(cur);
            cur = cur.next;
        }

        ListNode res = null;
        while (!stackA.isEmpty() && !stackB.isEmpty() && stackA.peek() == stackB.peek()) {
            res = stackA.pop();
            stackB.pop();
        }
        return res;
    }

    /**
     * 解题思路：O(1)的空间复杂度，思路如下：
     * 1 统计两个链表的长度，记为len1和len2
     * 2 然后让两个链表从同一长度起点开始往后走，比如长度5和7，那么让第二个链表也从index1（0-indexed）开始往后走，
     * 直到第一个node相同的位置
     *
     * 温馨提示：其实不需要直到长度也能解决这个问题（当然，遍历次数是不会减少的，还是需要重复遍历部分数据）
     * @param headA
     * @param headB
     * @return
     */
    public ListNode getIntersectionNode2(ListNode headA, ListNode headB) {
        if (null == headA || null == headB) {
            return null;
        }

        int len1 = 0, len2 = 0;
        ListNode curA = headA;
        while (null != curA) {
            len1++;
            curA = curA.next;
        }
        ListNode curB = headB;
        while (null != curB) {
            len2++;
            curB = curB.next;
        }

        curA = headA;
        curB = headB;
        //哪个长，就让哪个先走
        if (len1 > len2) {
            curA = advanceNode(curA, len1 - len2);
        } else if (len1 < len2) {
            curB = advanceNode(curB, len2 - len1);
        }

        //接下来两个一起开始往后面走，直到遇到第一个node相同的节点
        while (null != curA && null != curB && curA != curB) {
            curA = curA.next;
            curB = curB.next;
        }

        return curA;
    }

    /**
     * 把node往前推n步
     *
     * @param node
     * @param steps
     * @return
     */
    private ListNode advanceNode(ListNode node, int steps) {
        int i = 0;
        while (i < steps) {
            node = node.next;
            ++i;
        }
        return node;
    }
}

File: c++/linked_list/_160_IntersectionOfTwoLinkedLists.cc
//
// Created by ljt on 2026/6/7.
//

#include <iostream>
#include <map>
#include <string>
#include "linked_list_utils.h"
using namespace std;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    /**
     * 解题思路：获取两个链表的交点，注意，这道题要求O(m+n)的时间复杂度和O(1)的空间复杂度，如果不要求空间复杂度，这道题将会很简单，使用栈就可以实现
     * 解法1： 首先获取两个链表的长度，然后计算长度差，记为n，接下来长的链表先走n步，然后短的也可以走，这样子当两个节点第一次相等的时候就是最终的答案
     * 解法2：两个链表虽然各自的长度不一样，但是两个链表长度加起来肯定是固定的，那么，我们使用两个指针，第一个指针指向第一个链表，第二个指针指向第二个链表，然后同时移动两个指针，
     * 当第一个指针到末尾的时候开始遍历第二条链表，当第二个指针到末尾的时候开始遍历第一个链表，那么如果两个链表存在交点，那么当到交点的时候，两个指针走过的长度一定是一样的，如果没有交点，
     * 那么两个指针会一起到nullptr，很天才的想法
     * @param headA
     * @param headB
     * @return
     */
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        if (!(headA != nullptr && headB != nullptr)) {
            return nullptr;
        }
        // 接下来将开始计算两个链表的交点
        int sizeA = 0, sizeB = 0;
        ListNode *cur = headA;
        while (cur != nullptr) {
            cur = cur->next;
            sizeA++;
        }
        cur = headB;
        while (cur != nullptr) {
            cur = cur->next;
            sizeB++;
        }
        int gap = abs(sizeA - sizeB);
        ListNode *longList = headA;
        ListNode *shortList = headB;
        if (sizeB > sizeA) {
            longList = headB;
            shortList = headA;
        }

        // 长的先走n步
        int i = 0;
        while (i < gap) {
            longList = longList->next;
            i++;
        }

        // 接下来两个链表一起往前走
        while (longList != nullptr && shortList != nullptr) {
            if (longList == shortList) {
                return longList;
            }
            longList = longList->next;
            shortList = shortList->next;
        }
        return nullptr;
    }

    /**
     * 解法2： 两个指针一起遍历两个链表
     * @param headA
     * @param headB
     * @return
     */
    ListNode *getIntersectionNode2(ListNode *headA, ListNode *headB) {
        if (!(headA != nullptr && headB != nullptr)) {
            return nullptr;
        }
        ListNode *curA = headA;
        ListNode *curB = headB;
        while (curA != curB) {
            curA = curA != nullptr ? curA->next : headB;
            curB = curB != nullptr ? curB->next : headA;
        }
        return curA;
    }
};


int main() {
    Solution s;

    // 4,1,8,4,5       5,6,1,8,4,5
    std::vector<int> nums = {4, 1};
    std::vector<int> nums2 = {5, 6, 1};
    std::vector<int> common = {8, 4, 5};
    ListNode *head = genLinkedList(nums);
    ListNode *head2 = genLinkedList(nums2);
    ListNode *commonHead = genLinkedList(common);
    getTailOfLinkedList(head)->next = commonHead;
    getTailOfLinkedList(head2)->next = commonHead;
    auto result = s.getIntersectionNode(head, head2);
    if (result == nullptr) {
        printf("No Intersection!!!!\n");
    } else {
        printf("Intersection value: %d\n", result->val);
    }

    ListNode *result2 = s.getIntersectionNode2(head, head2);
    if (result2 == nullptr) {
        printf("No Intersection!!!!\n");
    } else {
        printf("Intersection value: %d\n", result2->val);
    }

    nums = {2, 6, 4};
    nums2 = {1, 5};
    head = genLinkedList(nums);
    head2 = genLinkedList(nums2);
    result = s.getIntersectionNode(head, head2);
    if (result == nullptr) {
        printf("No Intersection!!!!\n");
    } else {
        printf("Intersection value: %d\n", result->val);
    }

    result2 = s.getIntersectionNode2(head, head2);
    if (result2 == nullptr) {
        printf("No Intersection!!!!\n");
    } else {
        printf("Intersection value: %d\n", result2->val);
    }
}
